src/out_queue.rs

```diff
@@ -0,0 +1,309 @@
+// Copyright 2018 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under the MIT license <LICENSE-MIT
+// http://opensource.org/licenses/MIT> or the Modified BSD license <LICENSE-BSD
+// https://opensource.org/licenses/BSD-3-Clause>, at your option. This file may not be copied,
+// modified, or distributed except according to those terms. Please review the Licences for the
+// specific language governing permissions and limitations relating to use of the SAFE Network
+// Software.
+
+use std::collections::{BTreeMap, VecDeque};
+use std::time::Instant;
+use {Priority, SocketConfig};
+
+/// Socket outgoing message queue with priorities and message expiration.
+#[derive(Default)]
+pub struct OutQueue {
+    inner: BTreeMap<Priority, VecDeque<(Instant, Vec<u8>)>>,
+    conf: SocketConfig,
+}
+
+impl OutQueue {
+    /// Constructs empty write queue.
+    pub fn new(conf: SocketConfig) -> Self {
+        Self {
+            inner: Default::default(),
+            conf,
+        }
+    }
+
+    /// Push buffer to the queue with the given priority.
+    pub fn push(&mut self, buf: Vec<u8>, priority: Priority) {
+        self.push_at(Instant::now(), buf, priority);
+    }
+
+    /// Drop expired messages.
+    pub fn drop_expired(&mut self) -> usize {
```

Comment:
Should we call these `drop_invalid_queues()` as they also now drop empty in addition to expired queues ? Similarly [this one](https://github.com/maidsafe/socket-collection/pull/18/files#diff-7e7f52f73b837d4ae3b4e11838767573R76). Finally [where the actual predicate lies ](https://github.com/maidsafe/socket-collection/pull/18/files#diff-7e7f52f73b837d4ae3b4e11838767573R86) which dictates what counts as expired, we can put a doc comment there instead, what you [put here](https://github.com/maidsafe/socket-collection/pull/18/files#diff-7e7f52f73b837d4ae3b4e11838767573R75).